File: lib/src/internal/serializer/tcf_data_serializer.dart
import 'package:usercentrics_sdk/src/internal/serializer/consent_disclosure_object_serializer.dart';
import 'package:usercentrics_sdk/src/model/tcf_data.dart';

class TCFDataSerializer {
  static TCFData deserialize(dynamic value) {
    return TCFData(
        features: (value['features'] as List)
            .map((e) => _deserializeFeature(e))
            .toList(),
        purposes: (value['purposes'] as List)
            .map((e) => _deserializePurpose(e))
            .toList(),
        specialFeatures: (value['specialFeatures'] as List)
            .map((e) => _deserializeSpecialFeature(e))
            .toList(),
        specialPurposes: (value['specialPurposes'] as List)
            .map((e) => _deserializeSpecialPurpose(e))
            .toList(),
        stacks:
            (value['stacks'] as List).map((e) => _deserializeStack(e)).toList(),
        vendors: (value['vendors'] as List)
            .map((e) => _deserializeVendor(e))
            .toList(),
        tcString: value['tcString'],
        thirdPartyCount: value['thirdPartyCount']);
  }

  static TCFFeature _deserializeFeature(dynamic value) {
    return TCFFeature(
      purposeDescription: value['purposeDescription'],
      illustrations: value['illustrations']?.cast<String>(),
      id: value['id'],
      name: value['name'],
    );
  }

  static TCFPurpose _deserializePurpose(dynamic value) {
    return TCFPurpose(
      purposeDescription: value['purposeDescription'],
      illustrations: value['illustrations']?.cast<String>(),
      id: value['id'],
      name: value['name'],
      consent: value['consent'],
      isPartOfASelectedStack: value['isPartOfASelectedStack'],
      legitimateInterestConsent: value['legitimateInterestConsent'],
      showConsentToggle: value['showConsentToggle'],
      showLegitimateInterestToggle: value['showLegitimateInterestToggle'],
      stackId: value['stackId'],
      numberOfVendors: value['numberOfVendors'],
    );
  }

  static TCFSpecialFeature _deserializeSpecialFeature(dynamic value) {
    return TCFSpecialFeature(
      purposeDescription: value['purposeDescription'],
      illustrations: value['illustrations']?.cast<String>(),
      id: value['id'],
      name: value['name'],
      consent: value['consent'],
      isPartOfASelectedStack: value['isPartOfASelectedStack'],
      stackId: value['stackId'],
      showConsentToggle: value['showConsentToggle'],
    );
  }

  static TCFSpecialPurpose _deserializeSpecialPurpose(dynamic value) {
    return TCFSpecialPurpose(
      purposeDescription: value['purposeDescription'],
      illustrations: value['illustrations']?.cast<String>(),
      id: value['id'],
      name: value['name'],
    );
  }

  static TCFStack _deserializeStack(dynamic value) {
    return TCFStack(
      description: value['description'],
      id: value['id'],
      name: value['name'],
      purposeIds: value['purposeIds']?.cast<int>() ?? [],
      specialFeatureIds: value['specialFeatureIds']?.cast<int>() ?? [],
    );
  }

  static TCFVendor _deserializeVendor(dynamic value) {
    return TCFVendor(
        consent: value['consent'],
        features: value['features']?.cast<int>() ?? [],
        flexiblePurposes: value['flexiblePurposes']?.cast<int>() ?? [],
        id: value['id'],
        legitimateInterestConsent: value['legitimateInterestConsent'],
        legitimateInterestPurposes:
            value['legitimateInterestPurposes']?.cast<int>() ?? [],
        name: value['name'],
        policyUrl: value['policyUrl'],
        purposes: value['purposes']?.cast<int>() ?? [],
        specialFeatures: value['specialFeatures']?.cast<int>() ?? [],
        specialPurposes: value['specialPurposes']?.cast<int>() ?? [],
        showConsentToggle: value['showConsentToggle'],
        showLegitimateInterestToggle: value['showLegitimateInterestToggle'],
        cookieMaxAgeSeconds: value['cookieMaxAgeSeconds'],
        usesNonCookieAccess: value['usesNonCookieAccess'],
        deviceStorageDisclosureUrl: value['deviceStorageDisclosureUrl'],
        usesCookies: value['usesCookies'],
        cookieRefresh: value['cookieRefresh'],
        dataSharedOutsideEU: value['dataSharedOutsideEU'],
        dataCategories: value['dataCategories']?.cast<int>() ?? [],
        dataRetention: _deserializeDataRetention(value['dataRetention']),
        vendorUrls: _deserializeVendorUrlList(value['vendorUrls']),
        deviceStorage: ConsentDisclosureObjectSerializer.deserialize(
            value['deviceStorage']),
        restrictions: _deserializeRestrictions(value['restrictions']));
  }

  static DataRetention? _deserializeDataRetention(dynamic dataRetention) {
    if (dataRetention == null) {
      return null;
    }
    return DataRetention(
      stdRetention: dataRetention['stdRetention'],
      specialPurposes:
          _deserializeRetentionPeriod(dataRetention['specialPurposes']),
      purposes: _deserializeRetentionPeriod(dataRetention['purposes']),
    );
  }

  static RetentionPeriod _deserializeRetentionPeriod(dynamic retentionPeriod) {
    final Map<int, int> idAndPeriod = {};

    var map = (retentionPeriod as Map);
    for (var key in map.keys) {
      idAndPeriod[key] = map[key];
    }
    return RetentionPeriod(idAndPeriod: idAndPeriod);
  }

  static List<VendorUrl> _deserializeVendorUrlList(dynamic vendorUrls) {
    final List<VendorUrl> list = [];

    for (var i = 0; i < vendorUrls.length; i++) {
      list.add(_deserializeVendorUrl(vendorUrls[i]));
    }

    return list;
  }

  static VendorUrl _deserializeVendorUrl(dynamic value) {
    return VendorUrl(
        langId: value['langId'],
        privacy: value['privacy'],
        legIntClaim: value['legIntClaim']);
  }

  static List<TCFVendorRestriction> _deserializeRestrictions(
      dynamic restrictions) {
    final List<TCFVendorRestriction> list = [];

    for (var i = 0; i < restrictions.length; i++) {
      list.add(_deserializeTCFVendorRestriction(restrictions[i]));
    }
    return list;
  }

  static TCFVendorRestriction _deserializeTCFVendorRestriction(dynamic value) {
    return TCFVendorRestriction(
        purposeId: value['purposeId'],
        restrictionType:
            _deserialiazeRestrictionType(value['restrictionType'])!);
  }

  static RestrictionType? _deserialiazeRestrictionType(dynamic value) {
    switch (value) {
      case 'NOT_ALLOWED':
        return RestrictionType.notAllowed;
      case 'REQUIRE_CONSENT':
        return RestrictionType.requireConsent;
      case 'REQUIRE_LI':
        return RestrictionType.requireLi;
    }
    return null;
  }
}

File: ios/Classes/Serializer/TCFDataSerializer.swift
import Usercentrics

extension TCFData {
    func serialize()-> Any {
        return[
            "tcString" : tcString,
            "features" : features.map { $0.serialize() },
            "purposes" : purposes.map { $0.serialize() },
            "specialFeatures" : specialFeatures.map { $0.serialize() },
            "specialPurposes" : specialPurposes.map { $0.serialize() },
            "stacks" : stacks.map { $0.serialize() },
            "vendors" : vendors.map { $0.serialize() },
            "thirdPartyCount": thirdPartyCount
        ]
    }
}

extension TCFFeature {
    func serialize()-> Any {
        return[
            "purposeDescription" : purposeDescription,
            "illustrations" : illustrations,
            "id" : id,
            "name" : name,
        ]
    }
}

extension TCFPurpose {
    func serialize()-> Any {
        return[
            "purposeDescription" : purposeDescription,
            "illustrations" : illustrations,
            "id" : id,
            "name" : name,
            "consent" : Bool(from: self.consent) as Any,
            "isPartOfASelectedStack" : isPartOfASelectedStack,
            "legitimateInterestConsent" : Bool(from: self.legitimateInterestConsent) as Any,
            "showConsentToggle" : showConsentToggle,
            "showLegitimateInterestToggle" : showLegitimateInterestToggle,
            "stackId" : stackId as Any,
        ]
    }
}

extension TCFSpecialPurpose {
    func serialize()-> Any {
        return[
            "purposeDescription" : purposeDescription,
            "illustrations" : illustrations,
            "id" : id,
            "name" : name,
        ]
    }
}


extension TCFSpecialFeature {
    
    func serialize()-> Any {
        return[
            "purposeDescription" : purposeDescription,
            "illustrations" : illustrations,
            "id" : id,
            "name" : name,
            "consent" : Bool(from: self.consent) as Any ,
            "isPartOfASelectedStack" : isPartOfASelectedStack,
            "stackId" : stackId as Any,
            "showConsentToggle" : showConsentToggle,
        ]
    }
}

extension TCFStack {
    func serialize()-> Any {
        return [
            "description" : self.description,
            "id" : self.id,
            "name" : self.name,
            "purposeIds" : self.purposeIds,
            "specialFeatureIds" : self.specialFeatureIds,
        ]
    }
}

extension TCFVendor {
    func serialize()-> Any {
        return [
            "consent" : Bool(from: self.consent) as Any ,
            "features" : features.map { $0.id },
            "flexiblePurposes" : flexiblePurposes.map { $0.id },
            "id" : id,
            "legitimateInterestConsent" : Bool(from: self.legitimateInterestConsent) as Any ,
            "legitimateInterestPurposes" : legitimateInterestPurposes.map { $0.id },
            "name" : name,
            "policyUrl" : policyUrl,
            "purposes" : purposes.map { $0.id },
            "specialFeatures" : specialFeatures.map { $0.id },
            "specialPurposes" : specialPurposes.map { $0.id },
            "showConsentToggle" : showConsentToggle,
            "showLegitimateInterestToggle" : showLegitimateInterestToggle,
            "cookieMaxAgeSeconds" :  cookieMaxAgeSeconds as Any,
            "usesNonCookieAccess" : usesNonCookieAccess,
            "deviceStorageDisclosureUrl" : deviceStorageDisclosureUrl  as Any,
            "usesCookies" : usesCookies,
            "cookieRefresh" : Bool(from: self.cookieRefresh) as Any ,
            "dataSharedOutsideEU" : Bool(from: self.dataSharedOutsideEU) as Any,
            "dataRetention": (dataRetention?.serialize() ?? nil) as Any,
            "dataCategories" : dataCategories.map { $0.id },
            "vendorUrls" : vendorUrls.map { $0.serialize() },
            "deviceStorage" : self.deviceStorage?.serialize() as Any,
            "restrictions": self.restrictions.map { $0.serialize() }
        ]
    }
}

extension DataRetention {

    func serialize() -> Any {
        return [
            "stdRetention" : stdRetention?.intValue as? Any,
            "purposes" : purposes.idAndPeriod,
            "specialPurposes" : specialPurposes.idAndPeriod
        ]
    }
}

extension VendorUrl {

    func serialize() -> Any {
        return [
            "langId" : self.langId,
            "privacy" : self.privacy,
            "legIntClaim" : self.legIntClaim
        ]
    }
}

extension TCFVendorRestriction {

    func serialize() -> Any {
        return [
            "purposeId" : self.purposeId,
            "restrictionType" : self.restrictionType.name
        ] as [String : Any]
    }
}
